website/static/js/apiApplication.js

```diff
@@ -0,0 +1,434 @@
+            buttons:{
+                confirm:{
+                    label:'Deactivate',
+                    className:'btn-danger'
+                }
+            }
+        });
+    }
+});
+
+
+/*
+  ViewModel for Detail views (create and update pages- related though distinct behaviors in a single ViewModel)
+    Expects a urls object containing webListUrl, apiListUrl, and apiDetailUrl values. If apiDetailUrl is blank, it
+    behaves like a create view.
+ */
+var ApplicationDetailViewModel = oop.defclass({
+    constructor: function (urls) {
+        var placeholder = new ApplicationData();
+        this.appData = ko.observable(placeholder);
+
+        // Track whether data has changed, and whether user is allowed to leave page anyway
+        this.originalValues = ko.observable(placeholder.serialize());
+        this.dirty = ko.computed(function(){
+            return JSON.stringify(this.originalValues()) !== JSON.stringify(this.appData().serialize());
+        }.bind(this));
+        this.allowExit = ko.observable(false);
+
+        // Set up data access client
+        this.webListUrl = urls.webListUrl;
+        this.client = new ApplicationDataClient(urls.apiListUrl);
+
+        // Control success/failure messages above submit buttons
+        this.message = ko.observable();
+        this.messageClass = ko.observable();
+
+        // Toggle hiding client secret (in detail view)
+        this.showSecret = ko.observable(false);
+        // Toggle display of validation messages
+        this.showMessages = ko.observable(false);
+
+        // // If no detail url provided, render view as though it was a creation form. Otherwise, treat as READ/UPDATE.
+        this.apiDetailUrl = ko.observable(urls.apiDetailUrl);
+        this.isCreateView = ko.computed(function () {
+            return !this.apiDetailUrl();
+        }.bind(this));
+
+        // Prevent user from leaving page if there are unsaved changes
+        $(window).on('beforeunload', function () {
+            if (this.dirty() && !this.allowExit()) {
+                return 'There are unsaved changes on this page.';
+            }
+        }.bind(this));
+    },
+    init: function () {
+        if (!this.isCreateView()) {
+            var request = this.client.fetchOne(this.apiDetailUrl());
+            request.done(function (dataObj) {
+                this.appData(dataObj);
+                this.originalValues(dataObj.serialize());
+            }.bind(this));
+            request.fail(function(xhr, status, error) {
+                $osf.growl('Error',
+                             language.apiOauth2Application.dataFetchError,
+                            'danger');
+
+                Raven.captureMessage('Error fetching application data', {
+                    url: this.apiDetailUrl(),
+                    status: status,
+                    error: error
+                });
+            }.bind(this));
+        }
+    },
+    updateApplication: function () {
```

Comment:
Methods that do async work should return a promise. In this case, you can just return `request`.